### PR TITLE
apply rate limit to queue events

### PR DESCRIPTION
**Pull request by richardstartin**
# What Does This Do

Avoids creating huge numbers of queue time events in degenerate cases. Typically we see a few hundred of these per recording.

# Motivation

# Additional Notes

# Contributor Checklist

- [ ] Format the title [according the contribution guidelines](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#title-format)
- [ ] Assign the `type:` and (`comp:` or `inst:`) labels in addition to [any usefull labels](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#labels)
- [ ] Squash your commits prior merging or merge using GitHub's [Squash and merge](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#squash-and-merge-your-commits)
- [ ] Don't use `close`, `fix` or any [linking keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) when referencing an issue.  
  Use `solves` instead, and assign the PR [milestone](https://github.com/DataDog/dd-trace-java/milestones) to the issue
- [ ] Update the [public documentation](https://docs.datadoghq.com/tracing/trace_collection/library_config/java/) in case of new configuration flag or behavior

Jira ticket: [PROJ-IDENT]

### CODE REVIEW

File: internal-api/src/main/java/datadog/trace/api/sampling/PerRecordingRateLimiter.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        limit / samplingWindowsPerRecording(recordingLength.getSeconds(), windowDuration);
+    sampler = new AdaptiveSampler(windowDuration, samplesPerWindow, lookback, budgetLookback, true);
+  }
+
+  public boolean permit() {
+    if (!sampler.sample()) {
+      return false;
+    }
+    return true;
+    // return sampler.sample();
```
Comment (jbachorik) [inline internal-api/src/main/java/datadog/trace/api/sampling/PerRecordingRateLimiter.java:27]: nit: commented out code

File: internal-api/src/main/java/datadog/trace/api/sampling/PerRecordingRateLimiter.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    int samplesPerWindow =
+        limit / samplingWindowsPerRecording(recordingLength.getSeconds(), windowDuration);
+    sampler = new AdaptiveSampler(windowDuration, samplesPerWindow, lookback, budgetLookback, true);
+  }
+
+  public boolean permit() {
+    if (!sampler.sample()) {
+      return false;
+    }
+    return true;
```
Comment (jbachorik) [inline internal-api/src/main/java/datadog/trace/api/sampling/PerRecordingRateLimiter.java:26]: What is the difference vs. just doing `return sampler.sample()`?

**Comment by richardstartin** [comment]
Converting back to draft because this is going to require more work to support graal

**Comment by richardstartin** [comment]
@jbachorik you noticed some debugging artifacts, they've been removed.

**Review by jbachorik — APPROVED** [review]

**Pull request merged**